lib/ui/window/pointer_data_packet_converter.cc

```diff
@@ -0,0 +1,290 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/lib/ui/window/pointer_data_packet_converter.h"
+#include "flutter/fml/logging.h"
+
+#include <string.h>
+
+namespace flutter {
+
+PointerDataPacketConverter::PointerDataPacketConverter() : pointer_(0) {}
+
+PointerDataPacketConverter::~PointerDataPacketConverter() = default;
+
+std::unique_ptr<PointerDataPacket> PointerDataPacketConverter::Expand(
+    std::unique_ptr<PointerDataPacket> packet) {
+  size_t kBytesPerPointerData = kPointerDataFieldCount * kBytesPerField;
+  auto buffer = packet->data();
+  size_t buffer_length = buffer.size();
+
+  std::vector<PointerData> expended_pointers;
+  // Expands each pointer data in the buffer and stores it in the
+  // expended_pointers.
+  for (size_t i = 0; i < buffer_length / kBytesPerPointerData; i++) {
+    PointerData pointer_data;
+    memcpy(&pointer_data, &buffer[i * kBytesPerPointerData],
+           sizeof(PointerData));
+    ExpandPointerData(pointer_data, expended_pointers);
+  }
+
+  // Writes expended_pointers into expended_packet.
+  auto expended_packet =
+      std::make_unique<flutter::PointerDataPacket>(expended_pointers.size());
+  size_t count = 0;
+  for (auto& expended_pointer : expended_pointers) {
+    expended_packet->SetPointerData(count++, expended_pointer);
+  }
+  packet.reset();
+
+  return expended_packet;
+}
+
+void PointerDataPacketConverter::ExpandPointerData(
+    PointerData pointer_data,
+    std::vector<PointerData>& expended_pointers) {
+  // This method is responsible for freeing unused memory. If pointer_data will
+  // not be part of expended_pointers, it should free its memory.
+  if (pointer_data.signal_kind == PointerData::SignalKind::kNone) {
+    switch (pointer_data.change) {
+      case PointerData::Change::kCancel: {
+        // Android's three finger gesture will send a cancel event
+        // to a non-existing pointer. Drops the cancel if pointer
+        // is not previously added.
+        // https://github.com/flutter/flutter/issues/20517
+        auto iter = states_.find(pointer_data.device);
+        if (iter != states_.end()) {
+          PointerState state = iter->second;
+          FML_DCHECK(state.isDown);
```

Comment:
I am not sure if i should use static_assert or FML_DCHECK

Reply:
Static assertions cannot be used here because neither the isDown field or the state is known at at compile time. The DCHECK is the way to go. Consider adding a message that makes it clearer what to do when the assertion is tripped however.